rust/rope/src/find.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of string finding in ropes.
+
+use rope::RopeInfo;
+use tree::Cursor;
+
+use memchr::{memchr, memchr2, memchr3};
+
+pub enum FindResult {
+    Found(usize),
+    NotFound,
+    TryAgain,
+}
+
+/// Case-sensitive find (just does byte comparison, no Unicode logic).
+///
+/// Can panic if `pat` is empty.
+pub fn find(cursor: &mut Cursor<RopeInfo>, pat: &str) -> Option<usize> {
+    loop {
+        match find_step(cursor, pat) {
+            FindResult::Found(pos) => return Some(pos),
+            FindResult::NotFound => return None,
+            _ => (),
+        }
+    }
+}
+
+/// Case-insensitive find, using Rust's `to_lowercase` as the basis
+/// for comparison.
+///
+/// Can panic if `pat` is empty.
+
+// Note: one limitation of this method is that 'k' as the first
+// character in `pat` will not match against U+212A (kelvin). Seems
+// low-frequency enough, but should be fixed eventually.
+pub fn find_casei(cursor: &mut Cursor<RopeInfo>, pat: &str) -> Option<usize> {
+    let pat_lower = pat.to_lowercase();
+    let b = pat_lower.as_bytes()[0];
+    if b == b'i' {
+        loop {
+            match find_ascii_i_step(cursor, &pat_lower) {
+                FindResult::Found(pos) => return Some(pos),
+                FindResult::NotFound => return None,
+                _ => (),
+            }
+        }
+    } else if b >= b'a' && b <= b'z' {
+        loop {
+            match find_ascii_alpha_step(cursor, &pat_lower) {
+                FindResult::Found(pos) => return Some(pos),
+                FindResult::NotFound => return None,
+                _ => (),
+            }
+        }
+    } else if b < 0x80 {
+        loop {
+            match find_ascii_step(cursor, &pat_lower) {
+                FindResult::Found(pos) => return Some(pos),
+                FindResult::NotFound => return None,
+                _ => (),
+            }
+        }
+    } else {
+        loop {
+            match find_casei_step(cursor, &pat_lower) {
+                FindResult::Found(pos) => return Some(pos),
+                FindResult::NotFound => return None,
+                _ => (),
+            }
+        }
+    }
+}
+
+/// Case-sensitive find (just does byte comparison, no Unicode logic).
+///
+/// This method is like [`find`] but does a bounded amount of work.
```

Comment:
To have rustdoc generate the links, this should be (I believe) ``[`find`][find]``.